File: two pointers/3sum.py
def threeSum(nums: list[int]) -> list[list[int]]:
    nums.sort()
    ans = []

    for i, num in enumerate(nums):
        if num > 0:
            break
        if i > 0 and num == nums[i - 1]:
            continue
        l, r = i + 1, len(nums) - 1

        while l < r:
            left, right = nums[l], nums[r]
            added = num + left + right
            if added > 0:
                r -= 1
            elif added < 0:
                l += 1
            else:
                ans.append([num, left, right])
                l += 1
                r -= 1
                while nums[l] == nums[l - 1] and l < r:
                        l += 1

    return ans

print(threeSum([-1,0,1,2,-1,-4]))

File: two pointers/3sum.cs
public class SolutionThreeSum
{
    public IList<IList<int>> ThreeSum(int[] nums)
    {
        Array.Sort(nums);
        List<IList<int>> ans = new List<IList<int>>();
        HashSet<string> seen = new HashSet<string>();

        for (int i = 0; i < nums.Length - 2; ++i)
        {
            if (nums[i] > 0) break;
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            int l = i + 1;
            int r = nums.Length - 1;

            while (l < r)
            {
                int sum = nums[i] + nums[l] + nums[r];
                if (sum < 0) ++l;
                else if (sum > 0) --r;
                else
                {
                    List<int> triplet = new List<int> { nums[i], nums[l], nums[r] };
                    string key = string.Join(",", triplet);
                    if (!seen.Contains(key))
                    {
                        seen.Add(key);
                        ans.Add(triplet);
                    }
                    ++l;
                    --r;
                    while (l < r && nums[l] == nums[l - 1]) ++l;
                    while (l < r && nums[r] == nums[r + 1]) --r;
                }
            }
        }

        return ans;
    }
}
